components/ws-manager/pkg/manager/manager.go

```diff
@@ -217,6 +251,29 @@ func (m *Manager) StartWorkspace(ctx context.Context, req *api.StartWorkspaceReq
 	return okResponse, nil
 }
 
+func podRunning(clientset client.Client, podName, namespace string) wait.ConditionWithContextFunc {
+	return func(ctx context.Context) (bool, error) {
+		var pod corev1.Pod
+		err := clientset.Get(ctx, types.NamespacedName{Namespace: namespace, Name: podName}, &pod)
+		if err != nil {
+			return false, nil
+		}
+
+		switch pod.Status.Phase {
+		case corev1.PodFailed, corev1.PodSucceeded:
+			return false, fmt.Errorf("pod ran to completion")
+		case corev1.PodPending:
+			if pod.Status.Reason == "OutOfMemory" || pod.Status.Reason == "OutOfCpu" {
+				return false, xerrors.Errorf("cannot schedule pod, reason: %s", pod.Status.Reason)
+			}
+
+			return false, fmt.Errorf("pod ran to completion")
+		}
+
+		return true, nil
```

Comment:
This might be an edge case, but what if the pod goes into `unknown` state? Should we rather add an explicit case of `corev1.PodRunning`?